firestore-tos-extension/functions/src/index.ts

```diff
@@ -0,0 +1,239 @@
+import * as functions from "firebase-functions";
+import * as admin from "firebase-admin";
+import * as log from "./logs";
+import config from "./config";
+import { termsConverter, acknowledgementConverter } from "./converter";
+
+import { getEventarc } from "firebase-admin/eventarc";
+
+const eventChannel =
+  process.env.EVENTARC_CHANNEL &&
+  getEventarc().channel(process.env.EVENTARC_CHANNEL, {
+    allowedEventTypes: process.env.EXT_SELECTED_EVENTS,
+  });
+
+if (admin.apps.length === 0) {
+  admin.initializeApp();
+}
+
+const auth = admin.auth();
+const db = admin.firestore();
+
+export const acceptTerms = functions.handler.https.onCall(
+  async (data, context) => {
+    // Checking that the user is authenticated.
+    if (!context.auth) {
+      throw new functions.https.HttpsError(
+        "unauthenticated",
+        "No valid authentication token provided."
+      );
+    }
+
+    /** check if tosId has been provided  */
+    if (!data.tosId) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "No tosId provided."
+      );
+    }
+
+    if (data.acknowledged === null) {
+      throw new functions.https.HttpsError(
+        "invalid-argument",
+        "No acknowledgement value provided."
+      );
+    }
+
+    /** check tos documents */
+    const tosDoc = await db
+      .collection(config.collectionPath)
+      .doc("agreements")
+      .collection("tos")
+      .doc(data.tosId)
+      .withConverter(termsConverter)
+      .get();
+
+    /** Return if no agreement exists  */
+    if (!tosDoc || !tosDoc.exists) {
+      console.warn("tosDoc does not exist");
+      return;
+    }
+
+    const acknowledgementDates = {};
+
+    if (data.acknowledged) {
+      //@ts-ignore
+      acknowledgementDates.acknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    if (!data.acknowledged) {
+      //@ts-ignore
+      acknowledgementDates.unacknowledgedDate =
+        admin.firestore.FieldValue.serverTimestamp();
+    }
+
+    /** Set new acknowledgment */
+    const ack = {
+      ...tosDoc.data(),
+      acknowledged: data.acknowledged,
+      noticeType: data.noticeType,
+      ...acknowledgementDates,
+    };
+
+    /** Add Acknowledgment */
+    await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .doc(data.tosId)
+      .withConverter(acknowledgementConverter)
+      .set(
+        { ...ack },
+        {
+          merge: true,
+        }
+      );
+
+    /** Find current acknowledgements */
+    const acknowledgements = await db
+      .collection(config.collectionPath)
+      .doc("acknowledgements")
+      .collection(context.auth.uid)
+      .withConverter(acknowledgementConverter)
+      .get()
+      .then(($) => $.docs.map((doc) => doc.data()));
+
+    /** Set claims on user and return */
+    const claims = {};
+    claims[process.env.EXT_INSTANCE_ID] = acknowledgements;
```

Comment:
Since custom claims are very expensive, we should store the minimal amount data structure possible. Perhaps acknowledgements should just be an array of notice IDs that were acknowledged.